docs/proposals/20250325-distributed-inference.md

```diff
@@ -0,0 +1,173 @@
+- **Implementing distributed model tuning:** The scope of this proposal is limited to inference. Distributed tuning is not addressed.
+- **Introducing new runtimes or deployment mechanisms:** This proposal focuses on enhancing the existing vLLM runtime and StatefulSet deployment strategy, rather than introducing alternatives like LeaderWorkerSet.
+
+## Requirements
+
+### Workspace API Changes
+
+No API changes are needed for the workspace. The existing workspace API implicitly supports multi-node distributed inference, so users can continue using the same Workspace spec without any changes due to the following:
+
+- The number of nodes is specified via `.resource.count` in the Workspace specification:
+
+```yaml
+...
+resource:
+  count: 2
+  instanceType: "Standard_ND96asr_v4"
+...
+```
+
+For pre-provisioned nodes, users may define a list in the existing API `.resource.preferredNodes` to ensure inference deployment to specific nodes:
+
+```yaml
+...
+resource:
+  count: 2
+  preferredNodes:
+    - my-favorite-node-1
+    - my-favorite-node-2
+...
+```
+
+- [x] For models requiring more than one GPU, validate that `# GPUs/instance × workspace.resource.count ≥ Required # GPUs for a preset model`, and `GPU memory * # GPUs ≥ Required model memory`. This validation is performed in the API server when creating or updating a workspace. If the condition is not met, an error message will be returned to the user. This is already implemented in [`api/v1beta1/workspace_validation.go`](https://github.com/kaito-project/kaito/blob/1815428804593eaa94de0d6f78d82b53e85d0137/api/v1beta1/workspace_validation.go#L293-L380) and does not require any changes.
+
+> [!WARNING]
+> Kaito respects the user-defined `workspace.resource.count` and deploys the model across all specified nodes, even if the model could fit on fewer. Users should be aware that requesting more nodes than optimal for a model might introduce unnecessary communication overhead and potentially degrade performance, particularly without high-speed networking.
+
+### vLLM Runtime Parameters
+
+Flag additions to the vLLM base command is needed to support multi-node distributed inference. Per vLLM's [guidance](https://docs.vllm.ai/en/latest/serving/distributed_serving.html):
+
+- [x] `--tensor-parallel-size`: Set to the number of GPUs per node to configure tensor parallelism. This parameter determines how the model’s tensors are split across GPUs within a single node.
+- [ ] `--pipeline-parallel-size`: Set to the number of nodes to configure pipeline parallelism. This parameter determines how different model layers are sharded across nodes.
+
+vLLM uses [Ray](https://www.ray.io/) as the default framework to manage its distributed inference runtime. vLLM has provided a convenient script called [`multi-node-serving.sh`](https://github.com/vllm-project/vllm/blob/main/examples/online_serving/multi-node-serving.sh) to start the Ray server. The same script can be used by worker pods to join the Ray cluster.
```

Comment:
If we use a script `multi-node-serving.sh` to launch ray, should we consider how to maintain the ray application, like crash, restart, etc.

Reply:
Definitely. We are currently testing different failure scenarios and will update the proposal with our findings

cc @zhuangqh

Reply:
Addressed in https://github.com/kaito-project/kaito/pull/1105